File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Auth/PresenterViewController/PresenterViewController.m
//
//  RootParentVC.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 1/29/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "PresenterViewController.h"
#import "UsersViewController.h"
#import "AuthorizationViewController.h"
#import "UINavigationController+Appearance.h"
#import "Profile.h"

@interface PresenterViewController ()
//MARK: - Properties
@property (nonatomic, strong) UIViewController *current;
@end

@implementation PresenterViewController
//MARK: - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    Profile *profile = [[Profile alloc] init];
    profile.isFull ? [self showUsersScreen] : [self showLoginScreen];
}

// MARK: - Internal Methods
- (void)showLoginScreen {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Authorization" bundle:nil];
    AuthorizationViewController *authorizationVC = [storyboard instantiateViewControllerWithIdentifier:@"AuthorizationViewController"];
    UINavigationController *authNavVC = [[UINavigationController alloc] initWithRootViewController:authorizationVC];
    [authNavVC setupAppearanceWithColor:nil titleColor:UIColor.whiteColor];
    [authorizationVC setOnCompleteAuth:^(void) {
        [self showUsersScreen];
    }];
    if (self.current == nil) {
        self.current = authNavVC;
    }
    [self changeCurrentViewController:authNavVC];
}

- (void)showUsersScreen {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Users" bundle:nil];
    UsersViewController *usersScreen = [storyboard instantiateViewControllerWithIdentifier:@"UsersViewController"];
    UINavigationController *usersNavVC = [[UINavigationController alloc] initWithRootViewController:usersScreen];
    [usersNavVC setupAppearanceWithColor:nil titleColor:UIColor.whiteColor];
    [usersScreen setOnSignOut:^{
        [self showLoginScreen];
    }];
    if (self.current == nil) {
        self.current = usersNavVC;
    }
    [self changeCurrentViewController:usersNavVC];
}

//MARK - Setup
- (void)changeCurrentViewController:(UIViewController *)new {
    [self addChildViewController:new];
    new.view.frame = self.view.bounds;
    [self.view addSubview:new.view];
    [new didMoveToParentViewController:self];
    
    if ([self.current isEqual:new]) {
        return;
    }
    [self.current willMoveToParentViewController:nil];
    [self.current.view removeFromSuperview];
    [self.current removeFromParentViewController];
    self.current = new;
}

- (UIStatusBarStyle) preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Authorization/PresenterViewController.swift
//
//  PresenterViewController.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 1/27/20.
//  Copyright © 2020 quickBlox. All rights reserved.
//

import UIKit

class AuthNavigationController: UINavigationController { }
class UsersNavigationController: UINavigationController { }

class PresenterViewController: UIViewController {
    //MARK: - Properties
    private var current: UIViewController!
    private let profile = Profile()
    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        profile.isFull == false ? showLoginScreen() : showUsersScreen()
    }
    
    //MARK: - Internal Methods
    private func showLoginScreen() {
        guard let authVC = Screen.authViewController() else {
            return
        }
        authVC.onCompleteAuth = { [weak self] in
            self?.showUsersScreen()
        }
        let authNavVC = AuthNavigationController(rootViewController: authVC)
        authNavVC.navigationTitleColor(.white)
        if current == nil {
            current = authNavVC
        }
        changeCurrentViewController(authNavVC)
    }
    
    private func showUsersScreen() {
        guard let usersVC = Screen.usersViewController() else {
            return
        }
        let usersScreen = UsersNavigationController(rootViewController: usersVC)
        usersVC.onSignOut = { [weak self] in
            self?.showLoginScreen()
        }
        usersScreen.navigationTitleColor(.white)
        if current == nil {
            current = usersScreen
        }
        changeCurrentViewController(usersScreen)
    }
    
    private func changeCurrentViewController(_ newCurrentViewController: UIViewController) {
        addChild(newCurrentViewController)
        newCurrentViewController.view.frame = view.bounds
        view.addSubview(newCurrentViewController.view)
        newCurrentViewController.didMove(toParent: self)
        
        if current == newCurrentViewController {
            return
        }
        current.willMove(toParent: nil)
        current.view.removeFromSuperview()
        current.removeFromParent()
        current = newCurrentViewController
    }
    
    //MARK - Setup
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
}
